### ISSUE TITLE
Duplicate types in resulting BTF file

### ISSUE THREAD

**Issue by rafaeldtinoco** [kinvolk/libbpf#15]
When executing btfgen we are getting some type duplicates:

```
RELOCATION: [3289] struct in6_addr.in6_u.u6_addr8 (0:0:0 @ offset 0)
RELOCATION: [3289] struct in6_addr.in6_u.u6_addr8 (0:0:0 @ offset 0)
RELOCATION: [3294] struct ethhdr.h_proto (0:2 @ offset 12)
RELOCATION: [3289] struct in6_addr.in6_u.u6_addr16[5] (0:0:1:5 @ offset 10)
RELOCATION: [28129] struct __sk_buff.len (0:0 @ offset 0)
members for vfsmount is NULL. reloc_type 0x56367d1b8dd0. id 3073
members for file is NULL. reloc_type 0x56367d1b5d80. id 640
members for sockaddr_un is NULL. reloc_type 0x56367d254a00. id 49715
members for dentry is NULL. reloc_type 0x56367d20f0f0. id 954
done!
```

```
[34] STRUCT 'vfsmount' size=24 vlen=0
[53] STRUCT 'vfsmount' size=24 vlen=1
        'mnt_root' type_id=46 bits_offset=0
```

```
[44] STRUCT 'file' size=256 vlen=3
        'f_inode' type_id=22 bits_offset=256
        'f_flags' type_id=77 bits_offset=512
        'f_path' type_id=114 bits_offset=128
[72] STRUCT 'file' size=256 vlen=0
```

```
[11] STRUCT 'dentry' size=192 vlen=2
        'd_name' type_id=121 bits_offset=256
        'd_parent' type_id=46 bits_offset=192
[83] STRUCT 'dentry' size=192 vlen=0
```

**Comment by rafaeldtinoco** [comment]
First thing I noticed is that the 5.4.0-87 BTF file also has duplicated structs for dentry. They're pretty similar, with different BTF TYPE IDs and some members pointing to MOD types.

```
[954] STRUCT 'dentry' size=192 vlen=16
        'd_flags' type_id=9 bits_offset=0
        'd_seq' type_id=588 bits_offset=32
        'd_hash' type_id=940 bits_offset=64
        'd_parent' type_id=956 bits_offset=192
        'd_name' type_id=948 bits_offset=256
        'd_inode' type_id=959 bits_offset=384
        'd_iname' type_id=960 bits_offset=448
        'd_lockref' type_id=945 bits_offset=704
        'd_op' type_id=963 bits_offset=768
        'd_sb' type_id=965 bits_offset=832
        'd_time' type_id=1 bits_offset=896
        'd_fsdata' type_id=103 bits_offset=960
        '(anon)' type_id=951 bits_offset=1024
        'd_child' type_id=82 bits_offset=1152
        'd_subdirs' type_id=82 bits_offset=1280
        'd_u' type_id=953 bits_offset=1408

[28275] STRUCT 'dentry' size=192 vlen=16
        'd_flags' type_id=9 bits_offset=0
        'd_seq' type_id=588 bits_offset=32
        'd_hash' type_id=940 bits_offset=64
        'd_parent' type_id=28277 bits_offset=192
        'd_name' type_id=948 bits_offset=256
        'd_inode' type_id=28280 bits_offset=384
        'd_iname' type_id=960 bits_offset=448
        'd_lockref' type_id=945 bits_offset=704
        'd_op' type_id=28283 bits_offset=768
        'd_sb' type_id=28285 bits_offset=832
        'd_time' type_id=1 bits_offset=896
        'd_fsdata' type_id=103 bits_offset=960
        '(anon)' type_id=951 bits_offset=1024
        'd_child' type_id=82 bits_offset=1152
        'd_subdirs' type_id=82 bits_offset=1280
        'd_u' type_id=953 bits_offset=1408
```

This is expected and thats why libbpf relocation code speculates for multiple BTF types until it finds "res" (the resolution).

**Comment by rafaeldtinoco** [comment]
@mauriciovasquezbernal 

Alright, I think that what is going on is this: If, in function `bpf_reloc_info_get_btf`, I try to ignore the struct type if there are no members:

```C

		/* add members for struct and union */
		if (btf_is_struct(btf_type) || btf_is_union(btf_type)) {
			struct hashmap_entry *member_entry;
			struct btf_type *btf_type_cpy;
			int nmembers, new_size, bkt, index;

			if (reloc_type->members == NULL) {
				printf("members for %s is NULL. reloc_type %p. id %d\n",
					btf__str_by_offset(info->src_btf, btf_type->name_off), reloc_type, reloc_type->id);
			}

			nmembers = reloc_type->members ? hashmap__size(reloc_type->members) : 0;

			if (nmembers == 0)
				continue;
```

I get an error of: 

"error getting new id for old: 640. missing a type in ids_map ?"

And if we look for 640:

```
[640] STRUCT 'file' size=256 vlen=21
        'f_u' type_id=1175 bits_offset=0
        'f_path' type_id=988 bits_offset=128
        'f_inode' type_id=959 bits_offset=256
        'f_op' type_id=1158 bits_offset=320
        'f_lock' type_id=509 bits_offset=384
        'f_write_hint' type_id=1103 bits_offset=416
        'f_count' type_id=360 bits_offset=448
        'f_flags' type_id=9 bits_offset=512
        'f_mode' type_id=75 bits_offset=544
        'f_pos_lock' type_id=514 bits_offset=576
        'f_pos' type_id=66 bits_offset=832
        'f_owner' type_id=1173 bits_offset=896
        'f_cred' type_id=874 bits_offset=1152
        'f_ra' type_id=1174 bits_offset=1216
        'f_version' type_id=36 bits_offset=1472
        'f_security' type_id=103 bits_offset=1536
        'private_data' type_id=103 bits_offset=1600
        'f_ep_links' type_id=82 bits_offset=1664
        'f_tfile_llink' type_id=82 bits_offset=1792
        'f_mapping' type_id=622 bits_offset=1920
        'f_wb_err' type_id=1025 bits_offset=1984
```

the type is duplicated in the original BTF:

```
[28332] STRUCT 'file' size=256 vlen=21
        'f_u' type_id=1175 bits_offset=0
        'f_path' type_id=28307 bits_offset=128
        'f_inode' type_id=28280 bits_offset=256
        'f_op' type_id=28528 bits_offset=320
        'f_lock' type_id=509 bits_offset=384
        'f_write_hint' type_id=1103 bits_offset=416
        'f_count' type_id=360 bits_offset=448
        'f_flags' type_id=9 bits_offset=512
        'f_mode' type_id=75 bits_offset=544
        'f_pos_lock' type_id=514 bits_offset=576
        'f_pos' type_id=66 bits_offset=832
        'f_owner' type_id=28535 bits_offset=896
        'f_cred' type_id=28420 bits_offset=1152
        'f_ra' type_id=1174 bits_offset=1216
        'f_version' type_id=36 bits_offset=1472
        'f_security' type_id=103 bits_offset=1536
        'private_data' type_id=103 bits_offset=1600
        'f_ep_links' type_id=82 bits_offset=1664
        'f_tfile_llink' type_id=82 bits_offset=1792
        'f_mapping' type_id=28322 bits_offset=1920
        'f_wb_err' type_id=1025 bits_offset=1984
```

So, some relocations are made using type 640, others are using type 28332. They're the same, but some fields point to different types. Example:

```
'f_mapping' type_id=622 bits_offset=1920
[622] PTR '(anon)' type_id=621
[621] STRUCT 'address_space' size=160 vlen=15
...
```           

The STRUCT 'address_space' is also duplicated in original BTF. I think that is inherent to BTF deduplication algorithm:

```
'f_mapping' type_id=28322 bits_offset=1920
[28322] PTR '(anon)' type_id=28321
[28321] STRUCT 'address_space' size=160 vlen=15
...
```

But our algorithm does not take that into account.

**Comment by rafaeldtinoco** [comment]
We could do another pass and delete the empty struct. But then some types would point to inexistent types. Example:

The duplicate:
```
[37] STRUCT 'vfsmount' size=24 vlen=1
        'mnt_root' type_id=26 bits_offset=0
[38] STRUCT 'vfsmount' size=24 vlen=0
```

37 being used by anon pointer
```
[60] PTR '(anon)' type_id=37
```

38 being used by mount->mnt
```
[87] STRUCT 'mount' size=312 vlen=3
        'mnt_mountpoint' type_id=117 bits_offset=192
        'mnt_parent' type_id=43 bits_offset=128
        'mnt' type_id=38 bits_offset=256 
```

So, perhaps the correct thing to do would be:

1. To make all types pointing to empty struct to point to the other
2. To delete empty struct

Note: Unsure if having a "role" of type numbers in the BTF would be OK, or we would have to re-calculate all type ids to remove the one that was removed.

Note: Unsure if the empty struct in the BTF is satisfying relocations correctly. Maybe it is and all this is not needed.

WDYT ?

**Comment by mauriciovasquezbernal** [comment]
> First thing I noticed is that the 5.4.0-87 BTF file also has duplicated structs for dentry.

It's interesting. Perhaps btfgen is generating duplicated structures because they are already duplicated on the source BTF? I'll check it. 

> We could do another pass and delete the empty struct

I don't think we can do that. Please consider the following case. 

This C code (from tracee):

```
static __always_inline struct sockaddr_un get_unix_sock_addr(struct unix_sock *sock)
{
    struct unix_address* addr = READ_KERN(sock->addr);
    int len = READ_KERN(addr->len);
    struct sockaddr_un sockaddr = {};
    if (len <= sizeof(struct sockaddr_un)) {
        bpf_probe_read(&sockaddr, len, addr->name);
    }
    return sockaddr;
}
```
Generates a relocation like `RELOCATION: [49871] struct unix_address.name (0:3 @ offset 12)`.

Involved types are (original BTF):
```
[49870] STRUCT 'sockaddr_un' size=110 vlen=2
	'sun_family' type_id=3037 bits_offset=0
	'sun_path' type_id=8408 bits_offset=16
[49871] STRUCT 'unix_address' size=12 vlen=4
	'refcnt' type_id=790 bits_offset=0
	'len' type_id=21 bits_offset=32
	'hash' type_id=9 bits_offset=64
	'name' type_id=49872 bits_offset=96
[49872] ARRAY '(anon)' type_id=49870 index_type_id=21 nr_elems=0
```

Generated BTF is:
```
[45] STRUCT 'sockaddr_un' size=110 vlen=0
```

In this case it's correct to generate a struct without members. The code is trying to read the whole sockaddr_un structure, so we don't care about its members. 

> Note: Unsure if the empty struct in the BTF is satisfying relocations correctly. Maybe it is and all this is not needed.

Probably it's. I'll check it with more detail now.

**Comment by rafaeldtinoco** [comment]
> I don't think we can do that. Please consider the following case.
> 
> This C code (from tracee):
> 
> ```
> static __always_inline struct sockaddr_un get_unix_sock_addr(struct unix_sock *sock)
> {
>     struct unix_address* addr = READ_KERN(sock->addr);
>     int len = READ_KERN(addr->len);
>     struct sockaddr_un sockaddr = {};
>     if (len <= sizeof(struct sockaddr_un)) {
>         bpf_probe_read(&sockaddr, len, addr->name);
>     }
>     return sockaddr;
> }
> ```
> 
> Generates a relocation like `RELOCATION: [49871] struct unix_address.name (0:3 @ offset 12)`.
> 
> Involved types are (original BTF):
> 
> ```
> [49870] STRUCT 'sockaddr_un' size=110 vlen=2
> 	'sun_family' type_id=3037 bits_offset=0
> 	'sun_path' type_id=8408 bits_offset=16
> [49871] STRUCT 'unix_address' size=12 vlen=4
> 	'refcnt' type_id=790 bits_offset=0
> 	'len' type_id=21 bits_offset=32
> 	'hash' type_id=9 bits_offset=64
> 	'name' type_id=49872 bits_offset=96
> [49872] ARRAY '(anon)' type_id=49870 index_type_id=21 nr_elems=0
> ```
> 
> Generated BTF is:
> 
> ```
> [45] STRUCT 'sockaddr_un' size=110 vlen=0
> ```
> 
> In this case it's correct to generate a struct without members. The code is trying to read the whole sockaddr_un structure, so we don't care about its members.
> 

YES, I thought the exact same thing, about the same relocation (as this one is the only one with empty members and no duplicate).

**Comment by mauriciovasquezbernal** [comment]
TL;DR: It's because the source BTF has duplicated types.

We asked on the bpf mailing list and got this reply about the reason of duplicated types in the source BTF file: https://lore.kernel.org/all/CAHap4ztxPO485-5u5bkncyf9n-EQBTfF-3tN28jdNa4w1E-vkQ@mail.gmail.com/T/#u
